testtools/UART_interface/esp_arduino_uart_interface.py

```diff
@@ -0,0 +1,167 @@
+    # method to write to serial line with connection monitoring
+    def serial_write(self, ser, message, file=None):
+
+        # Check that the device is no longer sending bytes
+        if ser.in_waiting:
+            self.serial_read(ser, message, file)
+
+        # Check that the serial connection is open
+        if ser.writable():
+            wait = serial_settings.mxchip_buf_pause  # wait for at least 50ms between 128 byte writes.
+            buf = bytearray((message.strip() + '\r\n').encode('ascii'))
+            buf_len = len(buf)  # needed for loop as buf is a destructed list
+            bytes_written = 0
+            timeout = time.time()
+
+            while bytes_written < buf_len:
+                temp_written = ser.write(buf[:128])
+                buf = buf[temp_written:]
+                bytes_written += temp_written
+                # print("bytes written: %d" %bytes_written)
+                time.sleep(wait)
+                if (time.time() - timeout > serial_settings.serial_comm_timeout):
+                    break
+            # print("final written: %d" %bytes_written)
+            return bytes_written
+        else:
+            try:
+                time.sleep(2)
+                ser.open()
+                self.serial_write(ser, message, file)
+            except:
+                return False
+
+    # Read from serial line with connection monitoring
+    # If there is a sudden disconnect, program should report line in input script reached, and close files.
+    def serial_read(self, ser, message, file, first_read=False):
+
+        # Special per opt handling:
+        # if "send_telemetry" in message or "set_az_iothub" in message:
+        #     time.sleep(.15)
+        if "exit" in message and first_read:
+            time.sleep(serial_settings.wait_for_flash)
+            output = ser.in_waiting
+            while output < 4:
+                time.sleep(1)
+                output = ser.in_waiting
+                print("%d bytes in waiting" % output)
+
+        if ser.readable():
+            output = ser.readline(ser.in_waiting)
+            output = output.decode(encoding='utf-8', errors='ignore')
+
+            check_firmware_errors(output)
+            check_sdk_errors(output)
+            self.check_sample_errors(output)
+            print(output)
+            try:
+                # File must exist to write to it
+                file.writelines(output)
+            except:
+                pass
+
+            return output
+        else:
+            try:
+                time.sleep(2)
+                ser.open()
+                self.serial_read(ser, message, file, first_read=True)
+            except:
+                return False
+
+    # Note: the buffer size on the mxchip appears to be 128 Bytes.
+    def write_read(self, ser, input_file, output_file):
+        if 'esp32' in serial_settings.device_type:
+            serial_settings.bits_to_cache = 800
+            serial_settings.baud_rate = 1000000
```

Comment:
is this right? doesn't seem like most would be 115200

Reply:
the esp32 can do 1000000, and is set up to be on this baud on the sample.

Reply:
The esp32 can do 1000000, and the sample is set up to be on this baud for the esp32.